contracts/TokenSale.sol

```diff
@@ -310,6 +316,23 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
         }
     }
 
+    /**
+     * @dev withdraw funds for failed sales
+     */
+    function withdrawUserFunds() public {
+        require(hasEnded(), "Can only withdraw funds for ended sales!");
+        require(
+            !hasReachedSoftCap(),
+            "Can only withdraw funds for sales that didn't reach soft cap!"
+        );
+
+        uint256 investedEthRefund = ethInvestments[msg.sender];
```

Comment:
There is a reentrancy attack here. You should set the ethInvestments value back to zero